### ISSUE TITLE
Support multi-topic consumer subscriptions

### ISSUE THREAD

**Issue by ewencp** [confluentinc/kafka-rest#35]
Currently you can only subscribe to one topic because the old consumer only supports making one call to ConsumerConnector.createMessageStreams() and the REST proxy only subscribes to topics and calls that method when GET /consumer/{group}/instances/{instance}/topics/{topic} is called. It should also support specifying a set of topics up front in the initial consumer creation call to support multi-topic consumers. This should be optional and the old behavior should still work -- not specifying topics or even providing a null request entity body should still allow implicit subscriptions when the first topic is read.

**Comment by kjvalencik** [comment]
It it safe to make multiple concurrent requests to the same consumer but different topics?

**Comment by hachikuji** [comment]
@kjvalencik The old Kafka consumer has a method createMessageStreams which basically takes a list of topics and returns a separate handle (an iterator really) for consuming from each topic. These handles can be processed concurrently, but the problem is that you can only call createMessageStreams once, so you have to know all the topics that will be consumed. The new consumer (which is getting pretty close) has the ability to subscribe to new topics dynamically, which solves that problem. Unfortunately, it doesn't provide a convenient way to split the message streams by topic and it is not safe for concurrent access, so it will be an interesting challenge to make it suitable for this proxy.

**Comment by kjvalencik** [comment]
@hachikuji, thanks! Since you can only subscribe to one topic per consumer, is it safe to use the same groupId across multiple consumers each consuming different topics? Or should there be distinct groupIds for each topic?

**Comment by hachikuji** [comment]
@kjvalencik That's a good question. I think it would be safe to use the same groupId across multiple topics in this way. The downside is that since there are more instances in the same consumer group, the rebalances are going to be more costly. The nice thing about using separate groups is that consumers from one topic would be able to keep processing while the group from another topic is rebalancing. Since either way we are creating a new consumer instance internally, I'd probably recommend using separate groups. Once this ticket is resolved, however, it might be more efficient to reuse the same group since it would allow us to reuse the same consumer instance internally.

Just out of curiosity, can you describe your use case a little bit? Would you be satisfied with an API which allowed you to subscribe to multiple topics in the same group, but didn't provide a way to differentiate the message stream by topic? In other words, it would only give you a single endpoint to get messages from all of the subscribed topics. I ask because this is how the new consumer works.

**Comment by kjvalencik** [comment]
Thanks, that makes a lot of sense. The use case is we have single node.js processes (microservice architecture) that are consuming on multiple topics and doing separate things. I think the group per-topic will work well. The only reason I wanted to potentially share a group was that if in the future (with the new consumer API) it was more efficient to share a group, we wouldn't have to migrate our current comitted offsets.

A single stream to consume multiple topics could also work well for us in many cases. The only downside I can see is that committing offsets safely (e.g., after consumption) is a stop the world event, unlike with a single group per topic. However, I think in most of our use cases, the load is going to be amortized over a number of processes anyway and the extra latency incurred would be worth reducing operational overhead.

Thanks again for your quick answers.

**Comment by ewencp** [comment]
Agreed the new consumer will make it easier to implement some of these things. We could potentially support both modes (i.e. single topic by creating consumer then just reading `GET /consumer/{group}/instances/{instance}/topics/{topic}` and multi-topic mode by requiring more info in the initial creation request, then using a new path like `GET /consumer/{group}/instances/{instance}/messages`).

I the real decision here is whether it complicates things too much. We could document the API, but it could be confusing and we'd need to document some weird errors as well (e.g. "can't use topic resource when subscribed to multiple topics"). @hachikuji I'm specifically thinking it could end up as confusing as the new consumer API with both simple and high-level consumer modes.

**Comment by hachikuji** [comment]
@ewencp Another sort of lame option would be to try to use pause/resume in the new API to turn off all but one topic when we handle a request for that topic. In fact, that might not be too bad. We can allow the consumer to prefetch a certain amount of data from each topic and then pause them until that data has been consumed.

**Comment by ewencp** [comment]
@hachikuji Interesting idea. I was thinking we could also possibly use buffering (specific to the REST proxy) and then block fetch request if that buffer is full. ConsumerRecords could just be sorted into the right buffer as they are received. Buffering has the obvious drawback of adding more potential memory use and another option, but might be better if you are consistently consuming from both topics.

In fact, I guess with any buffering that can block consumption we'd have to use pause/resume as well because if a REST proxy buffer for a topic became full, we'd still need to be able to call `poll()` to get records for other topics if the application is waiting on a specific message from another topic.

The main drawback I see with only using the pause/resume is that the requests end up blocking each other -- if you have a client making requests for both topic partitions A-1 and topic B-1, if they have the same leader then we'll only send one fetch request at a time and we'll have to wait for data to be returned. This is the same issue we have with pause/resume for the normal Java client though, so maybe it can be resolved for this issue if we figure out how to address it in the Java client.

**Comment by kjvalencik** [comment]
If you start buffering and reading data from multiple topics wouldn't that cause internal offsets to get incremented before the data is actually read read? That could cause unexpected data loss when disabling `auto.commit` and doing things manually.

**Comment by ewencp** [comment]
@kjvalencik Correct, we'd need to do a bit more work to track the actual consumed offset if we introduced another layer of buffering.

**Comment by fangliuwh** [comment]
I get this post from this link: https://groups.google.com/forum/#!topic/confluent-platform/siQP0MBDHn8. have a very late comment on this ticket: tracking the offsets completely on the rest proxy server side would be very difficult task, given that we need to buffer data on rest proxy. What about leaving the offset commit to the client side of rest proxy. The committed offset is used to restart the consumer at a specified point; while the consumer, once started/initialized with the commit offset, it can keep its "temporal" offset (decoupled from the committed offset stored in kafka) and continuously consume the data until the buffer queue is full.  And the rest proxy client can continue to process the message and commit the offset accordingly. For this, we could have a new url like: POST consumer_group/offsets, with request body array of topic, partition, offset. Would this work?
Can I ask if you would have any plan on when to solve this issue? Thanks a lot!

**Comment by ewencp** [comment]
@fangliuwh https://github.com/confluentinc/kafka-rest/issues/107 is filed, but nobody has picked it up yet.

**Comment by kklipsch** [comment]
How does this issue jive with this quote documented here:

http://docs.confluent.io/3.0.0/kafka-rest/docs/api.html#get--consumers-(string-group_name)-instances-(string-instance)-topics-(string-topic_name)

"Error code 40901 – Consumer has already initiated a subscription. Consumers may subscribe to multiple topics, but all subscriptions must be initiated in a single request."
